# 墨尔本房价预测

## 将建立决策树和随机森林模型预测墨尔本市的房价。

## 第一步：决策树
1. 导入`pandas`库
2. 读取墨尔本市数据集
3. 预览数据集描述性统计摘要

In [1]:
import pandas as pd

In [3]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("dansbecker/melbourne-housing-snapshot")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/melbourne-housing-snapshot


In [6]:
melbourne_file_path = '/kaggle/input/melbourne-housing-snapshot/melb_data.csv'

melbourne_data = pd.read_csv(melbourne_file_path)

melbourne_data.describe()

,Rooms,Price,Distance,Postcode,Bedroom2,Bathroom,Car,Landsize,BuildingArea,YearBuilt,Lattitude,Longtitude,Propertycount
count,13580.000000,1.358000e+04,13580.000000,13580.000000,13580.000000,13580.000000,13518.000000,13580.000000,7130.000000,8205.000000,13580.000000,13580.000000,13580.000000
mean,2.937997,1.075684e+06,10.137776,3105.301915,2.914728,1.534242,1.610075,558.416127,151.967650,1964.684217,-37.809203,144.995216,7454.417378
std,0.955748,6.393107e+05,5.868725,90.676964,0.965921,0.691712,0.962634,3990.669241,541.014538,37.273762,0.079260,0.103916,4378.581772
min,1.000000,8.500000e+04,0.000000,3000.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1196.000000,-38.182550,144.431810,249.000000
25%,2.000000,6.500000e+05,6.100000,3044.000000,2.000000,1.000000,1.000000,177.000000,93.000000,1940.000000,-37.856822,144.929600,4380.000000
50%,3.000000,9.030000e+05,9.200000,3084.000000,3.000000,1.000000,2.000000,440.000000,126.000000,1970.000000,-37.802355,145.000100,6555.000000
75%,3.000000,1.330000e+06,13.000000,3148.000000,3.000000,2.000000,2.000000,651.000000,174.000000,1999.000000,-37.756400,145.058305,10331.000000
max,10.000000,9.000000e+06,48.100000,3977.000000,20.000000,8.000000,10.000000,433014.000000,44515.000000,2018.000000,-37.408530,145.526350,21650.000000


4. 预览数据集列名

In [7]:
melbourne_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

5. 删除有空值的行

In [8]:
melbourne_data = melbourne_data.dropna(axis=0)

6. 定义模型因变量`y`=价格`Price`

In [9]:
y = melbourne_data.Price

7. 把影响价格的主要因素列赋值给`melbourne_features`
8. 定义模型自变量`X`=`melbourne_features`

In [10]:
melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']
X = melbourne_data[melbourne_features]

9. 生成`X`变量述性统计摘要

In [11]:
X.describe()

,Rooms,Bathroom,Landsize,Lattitude,Longtitude
count,6196.000000,6196.000000,6196.000000,6196.000000,6196.000000
mean,2.931407,1.576340,471.006940,-37.807904,144.990201
std,0.971079,0.711362,897.449881,0.075850,0.099165
min,1.000000,1.000000,0.000000,-38.164920,144.542370
25%,2.000000,1.000000,152.000000,-37.855438,144.926198
50%,3.000000,1.000000,373.000000,-37.802250,144.995800
75%,4.000000,2.000000,628.000000,-37.758200,145.052700
max,8.000000,8.000000,37000.000000,-37.457090,145.526350


10. 展示`X`变量前五列

In [12]:
X.head()

,Rooms,Bathroom,Landsize,Lattitude,Longtitude
1,2,1.0,156.0,-37.8079,144.9934
2,3,2.0,134.0,-37.8093,144.9944
4,4,1.0,120.0,-37.8072,144.9941
6,3,2.0,245.0,-37.8024,144.9993
7,2,1.0,256.0,-37.8060,144.9954


11. 导入决策树库
12. 定义模型。为random_state指定一个数字，以确保每次运行都得到相同的结果
13. 拟合模型

In [13]:
from sklearn.tree import DecisionTreeRegressor

# 定义模型。为random_state指定一个数字，以确保每次运行都得到相同的结果
melbourne_model = DecisionTreeRegressor(random_state=1)

# 拟合模型
melbourne_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

14. 展示`X`变量前五列
15. 展示用`melbourne_model`模型预测`X`前五行变量的房价`y`变量`Price`

In [14]:
print("Making predictions for the following 5 houses:")
print(X.head())
print("The predictions are")
print(melbourne_model.predict(X.head()))

Making predictions for the following 5 houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The predictions are
[1035000. 1465000. 1600000. 1876000. 1636000.]


### 用平均绝对误差`MAE`验证决策树模型精度

16. 导入`sklearn.metrics`库`mean_absolute_error`模块
17. 验证预测`X`前五行变量的房价`y`变量`Price`的和原始数据集前五行房价对比的平均绝对误差

In [15]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melbourne_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

1115.7467183128902

数据解读：由结果可以看出用数据集前五行训练出的预测房价与数据集前五行真实房价差值是1115.7467183128902元

18. 数据分割：

使用 train_test_split 将数据集 (X, y) 划分为训练集和验证集

random_state=0 确保每次运行都能得到相同的随机分割结果（可复现性）

19. 模型定义与训练：

创建决策树回归模型 (DecisionTreeRegressor)

使用训练数据 (train_X, train_y) 拟合模型

20. 模型验证：

用训练好的模型对验证集 (val_X) 进行预测

计算预测值 (val_predictions) 与真实值 (val_y) 的平均绝对误差 (MAE)

打印出这个平均绝对误差`MAE`作为模型性能评估

In [16]:
from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every time we
# run this script.
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)
# Define model
melbourne_model = DecisionTreeRegressor()
# Fit model
melbourne_model.fit(train_X, train_y)

# get predicted prices on validation data
val_predictions = melbourne_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

272195.58209597593


21. 创建函数，通过调用函数来帮助比较`max_leaf_nodes`不同值的`MAE`分数

In [17]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

22. 添加实例`[5, 50, 500, 5000]`查看不同树的深度输出的`MAE`有什么不同，并选择最优数目深度

In [18]:
# compare MAE with differing values of max_leaf_nodes
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  385696
Max leaf nodes: 50  		 Mean Absolute Error:  279794
Max leaf nodes: 500  		 Mean Absolute Error:  261718
Max leaf nodes: 5000  		 Mean Absolute Error:  271320


数据解读：在列出的选项中，`500`的深度是最佳叶数。

## 第二步：随机森林

1. 用随机森林模型训练`train_test_split`将数据集

In [20]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

207190.6873773146


数据解读，输出`MAE`结果是`207190.6873773146`，相比起决策树最优深度`500`的`MAE`结果`261718`有显著改善。

2. 使用随机森林模型训练`melb_data.csv`墨尔本房价完整数据集

In [22]:
# To improve accuracy, create a new Random Forest model which you will train on all training data
rf_model_on_full_data = RandomForestRegressor(random_state=1)


# fit rf_model_on_full_data on all data from the training data
rf_model_on_full_data.fit(X, y)

RandomForestRegressor(random_state=1)

3. 导入需要预测的数据集

In [ ]:
# path to file you will use for predictions
test_data_path = '____'

# read test data file using pandas
test_data = pd.read_csv(test_data_path)

# create test_X which comes from test_data but includes only the columns you used for prediction.
# The list of columns is stored in a variable called features
test_X = test_data[melbourne_features]

# make predictions which we will submit.
test_preds = rf_model_on_full_data.predict(test_X)

4. 保存预测成果`ID`列及房价`SalePrice`列。另存为`csv`文件在根目录下

In [ ]:
# Run the code to save predictions in the format used for competition scoring

output = pd.DataFrame({'Id': test_data.Id,
                       'SalePrice': test_preds})
output.to_csv('submission.csv', index=False)